redhat-appstudio/service-provider-integration-operator
# Providing the option to specify dockerconfigjson auth format

### What does this PR do?
Kubernetes' interpretation of docker config.json is more permissive than Docker's. This PR allows users to alter the json we create for Secrets of type kubernetes.io/dockerconfigjson.

### Screenshot/screencast of this PR



### What issues does this PR fix or reference?

Implements https://issues.redhat.com/browse/SVPI-446


### How to test this PR?

1. Deploy and create binding 
```yaml
apiVersion: appstudio.redhat.com/v1beta1
kind: SPIAccessTokenBinding
metadata:
  name: test-binding-dockerconfigjson
  namespace: default
spec:
  permissions:
    required:
      - type: rw
        area: registry
  repoUrl: quay.io/repository/redhat_emp1/spi-test
  secret:
    type: kubernetes.io/dockerconfigjson
    name: test-secret-dockerconfigjson
    annotations:
      spi.appstudio.redhat.com/config-json-type: kubernetes
```
2. upload secret
3. check the secret created by the binding. The format should be as follows:
```json
{
    "auths": {
        "quay.io/repository/redhat_emp1/spi-test": {
            "auth": "…"
        }
    }
}
```
4. edit the secret's annotations in binding to:
```yaml
annotations:
      spi.appstudio.redhat.com/config-json-type: explicit
      spi.appstudio.redhat.com/config-json-auth-key: my.custom.host/test
```
and check the secret. It should have the format:
```json
{
    "auths": {
        "my.custom.host/test": {
            "auth": "…"
        }
    }
}
```
5. Edit the secret's annotations in binding to:
```yaml
annotations:
      spi.appstudio.redhat.com/config-json-type: docker
```
and check the secret. It should have the format:
```json
{
    "auths": {
        "quay.io": {
            "auth": "…"
        }
    }
}
```
6. Remove all secret annotations from the binding and ensure the format stays the same.

```
quay.io/redhat-appstudio/pull-request-builds:spi-controller-#
quay.io/redhat-appstudio/pull-request-builds:spi-oauth-#
```
